## Data Dictionarity

- Unnamed: 0: Timestamp sampled with period interval of 10 minutes.
- ActivePower: Active power generated by the windmill.
- AmbientTemperatue: Ambient temperature.
- BearingShaftTemperature: Bearing shaft temperature of the motor.
- Blade1PitchAngle: pitch angle of the wind turbine blade 1.
- Blade2PitchAngle: pitch angle of the wind turbine blade 2.
- Blade3PitchAngle: pitch angle of the wind turbine blade 3.
- ControlBoxTemperature: temperature of wind turbine control.
- GearboxBearingTemperature: Gearbox Bearing Temperature of the motor.
- GearboxOilTemperature: Gearbox Oil Temperature of the motor.
- GeneratorRPM: Measured RPM of the generator.
- GeneratorWinding1Temperature: Temperature of the winding in the generator 1.
- GeneratorWinding2Temperature: Temperature of the winding in the generator 2.
- HubTemperature: Temperature of Hub computer that control pitching of blades.
- MainBoxTemperature: Temperature of main box.
- NacellePosition: Position of the nacelle.
- ReactivePower: Reactive power generated by the windmill.
- RotorRPM: RPM of the rotor.
- TurbineStatus: Status of the turbine.
- WTG: Indicates that mechanical energy is instead converted to electricity, only G01.
- WindDirection: Wind direction in degrees. 
- WindSpeed: Wind speed in km/h.


## 0.0 Imports

In [18]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

### 0.1 Helper Functions

In [19]:
# Function that converts camel style to snake style
def camel_to_snake(s):
    return ''.join(['_'+c.lower() if c.isupper() else c for c in s]).lstrip('_')

In [20]:
# Function that calculate  the descriptive statistics
def descriptive_analysis(numerical_df):
    # Function that calculate all the statistics measures for the dataset
    # Central tendency - Mean, median
    ct1 = pd.DataFrame(numerical_df.apply(np.mean)).T
    ct2 = pd.DataFrame(numerical_df.apply(np.median)).T

    # Dispersion - std, min, max, range, skew, kurtosis
    d1 = pd.DataFrame(numerical_df.apply(np.std)).T
    d2 = pd.DataFrame(numerical_df.apply(min)).T
    d3 = pd.DataFrame(numerical_df.apply(max)).T
    d4 = pd.DataFrame(numerical_df.apply(lambda x: x.max() - x.min())).T
    d5 = pd.DataFrame(numerical_df.apply(lambda x: x.skew())).T
    d6 = pd.DataFrame(numerical_df.apply(lambda x: x.kurtosis())).T

    # Concatenate
    m = pd.concat([ct1, ct2, d1,d2,d3,d4,d5,d6]).T.reset_index()
    m.columns = ["Attributes", "Mean", "Median", "Std", "Min", "Max", "Range", "Skew", "Kurtosis"]

    return m

### 0.2 Loading Data

In [21]:
df0 = pd.read_csv('../data/Turbine_Data.csv')

In [22]:
df0.head()

Unnamed: 0  ActivePower  AmbientTemperatue  \
0  2017-12-31 00:00:00+00:00          NaN                NaN   
1  2017-12-31 00:10:00+00:00          NaN                NaN   
2  2017-12-31 00:20:00+00:00          NaN                NaN   
3  2017-12-31 00:30:00+00:00          NaN                NaN   
4  2017-12-31 00:40:00+00:00          NaN                NaN   

   BearingShaftTemperature  Blade1PitchAngle  Blade2PitchAngle  \
0                      NaN               NaN               NaN   
1                      NaN               NaN               NaN   
2                      NaN               NaN               NaN   
3                      NaN               NaN               NaN   
4                      NaN               NaN               NaN   

   Blade3PitchAngle  ControlBoxTemperature  GearboxBearingTemperature  \
0               NaN                    NaN                        NaN   
1               NaN                    NaN                        NaN   
2               NaN                    NaN                        NaN   
3               NaN                    NaN                        NaN   
4               NaN                    NaN                        NaN   

   GearboxOilTemperature  ...  GeneratorWinding2Temperature  HubTemperature  \
0                    NaN  ...                           NaN             NaN   
1                    NaN  ...                           NaN             NaN   
2                    NaN  ...                           NaN             NaN   
3                    NaN  ...                           NaN             NaN   
4                    NaN  ...                           NaN             NaN   

   MainBoxTemperature  NacellePosition  ReactivePower  RotorRPM  \
0                 NaN              NaN            NaN       NaN   
1                 NaN              NaN            NaN       NaN   
2                 NaN              NaN            NaN       NaN   
3                 NaN              NaN            NaN       NaN   
4                 NaN              NaN            NaN       NaN   

   TurbineStatus  WTG  WindDirection WindSpeed  
0            NaN  G01            NaN       NaN  
1            NaN  G01            NaN       NaN  
2            NaN  G01            NaN       NaN  
3            NaN  G01            NaN       NaN  
4            NaN  G01            NaN       NaN  

[5 rows x 22 columns]

## 1.0 Data Description

Describing data is important in this section for better comprehension of the dataset. Also, by using Descriptive Statistics,  it is possible to describe data through methods such as graphical representations, measures of central tendency and measures of variability. It summarizes the data in a meaningful way which enables us to generate insights from it.

In [23]:
df1 = df0.copy()

### 1.1 Rename Columns

In [24]:
# Rename timestamp column 
df1 = df1.rename(columns={'Unnamed: 0': 'Date'})

In [25]:
# Convert the camel case to snake case
cols_old = ['Date', 'ActivePower', 'AmbientTemperatue', 'BearingShaftTemperature','Blade1PitchAngle', 'Blade2PitchAngle', 
            'Blade3PitchAngle', 'ControlBoxTemperature', 'GearboxBearingTemperature','GearboxOilTemperature', 
            'GeneratorRPM', 'GeneratorWinding1Temperature', 'GeneratorWinding2Temperature', 'HubTemperature',
            'MainBoxTemperature','NacellePosition', 'ReactivePower', 'RotorRPM', 'TurbineStatus', 'WTG',
            'WindDirection', 'WindSpeed']

snake_case = lambda x: camel_to_snake(x)

cols_new = list(map(snake_case, cols_old))

df1.columns = cols_new

# Correction after snake transformation
cols_to_correct = ['generator_r_p_m', 'rotor_r_p_m', 'w_t_g']
cols_corrected = ['generator_rpm', 'rotor_rpm', 'wtg']

df1 = df1.rename(columns={'generator_r_p_m': 'generator_rpm',
                         'rotor_r_p_m': 'rotor_rpm',
                         'w_t_g': 'wtg'})

### 1.2 Data Dimensions

In [26]:
# Print the number of examples and features
print("Number of rows: {}".format(df1.shape[0]))
print("Number of columns: {}".format(df1.shape[1]))

Number of rows: 118224
Number of columns: 22


### 1.3 Data Types

In [27]:
# Show the type of each column
df1.dtypes

date                               object
active_power                      float64
ambient_temperatue                float64
bearing_shaft_temperature         float64
blade1_pitch_angle                float64
blade2_pitch_angle                float64
blade3_pitch_angle                float64
control_box_temperature           float64
gearbox_bearing_temperature       float64
gearbox_oil_temperature           float64
generator_rpm                     float64
generator_winding1_temperature    float64
generator_winding2_temperature    float64
hub_temperature                   float64
main_box_temperature              float64
nacelle_position                  float64
reactive_power                    float64
rotor_rpm                         float64
turbine_status                    float64
wtg                                object
wind_direction                    float64
wind_speed                        float64
dtype: object

**Observations**:
- date is a object and not datetime.

In [28]:
# Convert date column from string to datetime
df1['date'] = pd.to_datetime(df1['date'])

### 1.4 Checking Missing Values

In [29]:
100* df1.isna().sum() / len(df1)

date                               0.000000
active_power                      19.855528
ambient_temperatue                20.644708
bearing_shaft_temperature         47.119028
blade1_pitch_angle                64.477602
blade2_pitch_angle                64.566416
blade3_pitch_angle                64.566416
control_box_temperature           47.421843
gearbox_bearing_temperature       47.100420
gearbox_oil_temperature           47.186696
generator_rpm                     47.307653
generator_winding1_temperature    47.196001
generator_winding2_temperature    47.177392
hub_temperature                   47.213764
main_box_temperature              47.128333
nacelle_position                  38.863513
reactive_power                    19.857220
rotor_rpm                         47.449756
turbine_status                    46.789146
wtg                                0.000000
wind_direction                    38.863513
wind_speed                        19.986636
dtype: float64

**Observations**:
- It was found a large number of missing values, representing 64.56% of the dataset in some features;
- There are some strategies that we can handle: Drop the columns with expressive missing values; Replace it by the mean/median; Moving average; Replace with the last or next value non-null.
- For this problem, I am going to use the method ffill and bfill from pandas, which propagates the last non-null value in backward and forward direction.


### 1.5 Replace NaN values

In [30]:
df1 = df1.fillna(method='ffill').fillna(method='bfill')

In [31]:
100* df1.isna().sum() / len(df1)

date                              0.0
active_power                      0.0
ambient_temperatue                0.0
bearing_shaft_temperature         0.0
blade1_pitch_angle                0.0
blade2_pitch_angle                0.0
blade3_pitch_angle                0.0
control_box_temperature           0.0
gearbox_bearing_temperature       0.0
gearbox_oil_temperature           0.0
generator_rpm                     0.0
generator_winding1_temperature    0.0
generator_winding2_temperature    0.0
hub_temperature                   0.0
main_box_temperature              0.0
nacelle_position                  0.0
reactive_power                    0.0
rotor_rpm                         0.0
turbine_status                    0.0
wtg                               0.0
wind_direction                    0.0
wind_speed                        0.0
dtype: float64

### 1.6 Descriptive Statistics

In [32]:
num_attributes = df1.select_dtypes(include = ['float64'])
cat_attributes = df1.select_dtypes(exclude = ['float64', 'datetime64[ns]'])

#### 1.6.1 Numerical Attributes

In [33]:
descriptive_analysis(num_attributes)

Attributes         Mean       Median            Std  \
0                     active_power   639.934548   423.445880     630.141658   
1               ambient_temperatue    28.574250    28.126315       4.308102   
2        bearing_shaft_temperature    44.966830    47.901936       6.404004   
3               blade1_pitch_angle    31.797893    45.736893      22.150311   
4               blade2_pitch_angle    30.718228    43.699357      21.173457   
5               blade3_pitch_angle    30.718228    43.699357      21.173457   
6          control_box_temperature     0.000000     0.000000       0.000000   
7      gearbox_bearing_temperature    68.885245    75.066281      11.675754   
8          gearbox_oil_temperature    60.146597    63.304417       5.812920   
9                    generator_rpm  1359.868503  1691.274462     527.575804   
10  generator_winding1_temperature    87.239668    98.608170      25.175065   
11  generator_winding2_temperature    86.582197    97.978180      25.170356   
12                 hub_temperature    38.700162    39.996185       5.820179   
13            main_box_temperature    40.595611    41.898426       5.925468   
14                nacelle_position   201.719506   188.000000      90.954493   
15                  reactive_power    84.387076    28.384563     116.504803   
16                       rotor_rpm    12.199976    15.160864       4.737832   
17                  turbine_status  1239.298501     2.000000  261586.939523   
18                  wind_direction   201.719506   188.000000      90.954493   
19                      wind_speed     6.031632     5.672165       2.819615   

           Min           Max         Range        Skew      Kurtosis  
0   -38.524659  1.779032e+03  1.817557e+03    0.648454     -1.082583  
1     0.000000  4.240560e+01  4.240560e+01    0.281966      0.811206  
2     0.000000  5.508866e+01  5.508866e+01   -3.534934     20.653067  
3   -43.156734  9.014361e+01  1.333003e+02   -0.436169     -0.989127  
4   -26.443415  9.001783e+01  1.164612e+02   -0.366678     -0.828616  
5   -26.443415  9.001783e+01  1.164612e+02   -0.366678     -0.828616  
6     0.000000  0.000000e+00  0.000000e+00    0.000000      0.000000  
7     0.000000  8.223793e+01  8.223793e+01   -2.404865      9.730579  
8     0.000000  7.076458e+01  7.076458e+01   -1.396449      3.684245  
9     0.000000  1.809942e+03  1.809942e+03   -1.351028      0.830080  
10    0.000000  1.267730e+02  1.267730e+02   -0.317960     -1.399759  
11    0.000000  1.260430e+02  1.260430e+02   -0.315514     -1.411226  
12    0.000000  4.799618e+01  4.799618e+01   -3.244384     17.797759  
13    0.000000  5.425000e+01  5.425000e+01   -3.126101     18.515855  
14    0.000000  3.570000e+02  3.570000e+02    0.009001     -0.872459  
15 -203.182591  4.037136e+02  6.068962e+02    1.183835      0.061266  
16    0.000000  1.627350e+01  1.627350e+01   -1.355248      0.851473  
17    0.000000  6.574653e+07  6.574653e+07  243.544426  59384.297691  
18    0.000000  3.570000e+02  3.570000e+02    0.009001     -0.872459  
19    0.000000  2.297089e+01  2.297089e+01    0.839137      0.972560

**Observations**:

- **active_power** has an average of 639.93 KW and it reaches a maximum value of 1.78 MW. The variable has low skew and kurtosis, it may follow a normal distribution.
- **ambient_temperature** may follow a normal distribution and the average value is 28.56 ºC.
- **bearing_shaft_temperature** has a range of 55 ºC, and its average is 44.96 ºC.
- All the **blade** features are similar, in which the blade 1 vary a little bit more than the others.
- **control_box_temperature** has no information, all the values are zero.
- **gearbox_bearing_temperature** is skewed compared to the other variables, and this may not follow a normal distribution. The range is 82.2 ºC and the median 75.06 ºC.
- **gearbox_oil_temperature** may follow the same distribution of gear box bearing temperature. 
- The maximum value of **generator_rpm** is 1809 RPM, while its minimum is zero (when it is off). The median value is 1691 RPM.
- **generator_winding_temperature 1 and 2** are similar, they may not follow a normal distribution and the maximum values reach 1260 ºC, which can indicate an outlier. 
- **hub_temperature** reaches 47.96 ºC with an average of 38.7 ºC.
- **main_box_temperature** is very similar to the hub temperature, but it reaches 54.25 ºC.
- **nacelle_position** has a range of degrees from 0 to 357º, with a median of 188º and an average of 201.71º.
- **reactive_power** has a minimum value of -203.18 VAr, which indicates a moment that the load is more capacitive. When the reactive power is positive, it means a inductive load. Mostly values are positive, it indicates a inductive load.
- **rotor_rpm** is very low compared to the generator. The average value is 12.19 RPM.
- **wind_direction** also has a range from 0 to 357º and the average value is 201.71º. 
- **wind_speed** may follow a normal distribution and the average value is 6.03 km/h, reaching the maximum value of 22.9 km/h.

#### 1.6.2 Categorical Attributes

- Only WTG is a categorical attribute and has no information contained in the variable. I will treat it later.

## 2.0 Feature Engineering